File: L15GoldmanSachsHard/ShortestSubarrayWithSumAtLeastK.java

import java.util.ArrayDeque;
import java.util.Deque;

public class ShortestSubarrayWithSumAtLeastK {

    public int shortestSubarray(int[] nums, int k) {
        int n = nums.length;
        long[] prefixSum = new long[n + 1];

        // Compute prefix sums
        for (int i = 0; i < n; i++) {
            prefixSum[i + 1] = prefixSum[i] + nums[i];
        }

        int result = Integer.MAX_VALUE;
        Deque<Integer> deque = new ArrayDeque<>();

        for (int i = 0; i <= n; i++) {
            // Remove indices from the front of the deque if the sum is >= k
            while (!deque.isEmpty() && prefixSum[i] - prefixSum[deque.peekFirst()] >= k) {
                result = Math.min(result, i - deque.pollFirst());
            }

            // Maintain the deque in increasing order of prefix sums
            while (!deque.isEmpty() && prefixSum[i] <= prefixSum[deque.peekLast()]) {
                deque.pollLast();
            }

            deque.offerLast(i);
        }

        return result == Integer.MAX_VALUE ? -1 : result;
    }

    public static void main(String[] args) {
        ShortestSubarrayWithSumAtLeastK solution = new ShortestSubarrayWithSumAtLeastK();
        int[] nums = {2, -1, 2};
        int k = 3;
        System.out.println(solution.shortestSubarray(nums, k)); // Output: 3
    }
}

File: L15GoldmanSachsHard/ShortestSubarrayWithSumAtLeastK.cpp
#include<bits/stdc++.h>
#include <iostream>
#include <vector>
#include <deque>
#include <climits>

using namespace std;

int shortestSubarray(vector<int> &nums, int k)
{
    int n = nums.size();
    vector<long> prefixSum(n + 1, 0);

    // Compute prefix sums
    for (int i = 0; i < n; i++)
    {
        prefixSum[i + 1] = prefixSum[i] + nums[i];
    }

    int result = INT_MAX;
    deque<int> dq;

    for (int i = 0; i <= n; i++)
    {
        // Remove indices from the front of the deque if the sum is >= k
        while (!dq.empty() && prefixSum[i] - prefixSum[dq.front()] >= k)
        {
            result = min(result, i - dq.front());
            dq.pop_front();
        }

        // Maintain the deque in increasing order of prefix sums
        while (!dq.empty() && prefixSum[i] <= prefixSum[dq.back()])
        {
            dq.pop_back();
        }

        dq.push_back(i);
    }

    return result == INT_MAX ? -1 : result;
}

int main()
{
    vector<int> nums = {2, -1, 2};
    int k = 3;
    cout << shortestSubarray(nums, k) << endl; // Output: 3
    return 0;
}
